avalanche/training/supervised/supervised_contrastive_replay.py

```diff
@@ -0,0 +1,180 @@
+from typing import Sequence, Optional
+
+import torch
+from avalanche.training.plugins.evaluation import default_evaluator
+from avalanche.training.templates import SupervisedTemplate
+from torch.nn import Module
+from torch.optim import Optimizer
+from torch.utils.data import DataLoader
+from avalanche.core import BaseSGDPlugin
+from torchvision.transforms import Compose
+from torchvision.transforms import Lambda
+from avalanche.training.plugins import ReplayPlugin
+from avalanche.training.losses import SCRLoss
+from avalanche.training.storage_policy import ClassBalancedBuffer
+
+
+class SCR(SupervisedTemplate):
+    """
+    Supervised Contrastive Replay from https://arxiv.org/pdf/2103.13885.pdf.
+    This strategy trains an encoder network in a self-supervised manner to
+    cluster together examples of the same class while pushing away examples
+    of different classes. It uses the Nearest Class Mean classifier on the
+    embeddings produced by the encoder.
+
+    Accuracy cannot be monitored during training (no NCM classifier).
+    During training, NCRLoss is monitored, while during eval
+    CrossEntropyLoss is monitored.
+
+    The original paper uses an additional fine-tuning phase on the buffer
+    at the end of each experience (called review trick, but not mentioned
+    in the paper). This implementation does not implement the review trick.
+    """
+    def __init__(self,
+                 model: Module,
```

Comment:
fix type? should be `SCRModel`

Reply:
actually, this should also be checked explicitly, because otherwise the method doesn't work. Raising a `ValueError` may be the best choice here.

Reply:
Agreed